src-self-hosted/link.zig

```diff
@@ -89,447 +115,570 @@ pub fn openBinFile(allocator: *Allocator, file: fs.File, options: Options) !ElfF
-        /// Points to the previous and next neighbors, based on the `text_offset`.
-        /// This can be used to find, for example, the capacity of this `TextBlock`.
-        prev: ?*TextBlock,
-        next: ?*TextBlock,
-
-        pub const empty = TextBlock{
-            .local_sym_index = 0,
-            .offset_table_index = undefined,
-            .prev = null,
-            .next = null,
+    pub fn makeWritable(base: *File, dir: fs.Dir, sub_path: []const u8) !void {
+        try switch (base.tag) {
+            .Elf => @fieldParentPtr(Elf, "base", base).makeWritable(dir, sub_path),
+            .C => @fieldParentPtr(C, "base", base).makeWritable(dir, sub_path),
+            else => unreachable,
         };
+    }
 
-        /// Returns how much room there is to grow in virtual address space.
-        /// File offset relocation happens transparently, so it is not included in
-        /// this calculation.
-        fn capacity(self: TextBlock, elf_file: ElfFile) u64 {
-            const self_sym = elf_file.local_symbols.items[self.local_sym_index];
-            if (self.next) |next| {
-                const next_sym = elf_file.local_symbols.items[next.local_sym_index];
-                return next_sym.st_value - self_sym.st_value;
-            } else {
-                // We are the last block. The capacity is limited only by virtual address space.
-                return std.math.maxInt(u32) - self_sym.st_value;
-            }
-        }
+    pub fn makeExecutable(base: *File) !void {
+        try switch (base.tag) {
+            .Elf => @fieldParentPtr(Elf, "base", base).makeExecutable(),
+            else => unreachable,
+        };
+    }
 
-        fn freeListEligible(self: TextBlock, elf_file: ElfFile) bool {
-            // No need to keep a free list node for the last block.
-            const next = self.next orelse return false;
-            const self_sym = elf_file.local_symbols.items[self.local_sym_index];
-            const next_sym = elf_file.local_symbols.items[next.local_sym_index];
-            const cap = next_sym.st_value - self_sym.st_value;
-            const ideal_cap = self_sym.st_size * alloc_num / alloc_den;
-            if (cap <= ideal_cap) return false;
-            const surplus = cap - ideal_cap;
-            return surplus >= min_text_capacity;
-        }
-    };
+    pub fn updateDecl(base: *File, module: *Module, decl: *Module.Decl) !void {
+        try switch (base.tag) {
+            .Elf => @fieldParentPtr(Elf, "base", base).updateDecl(module, decl),
+            .C => @fieldParentPtr(C, "base", base).updateDecl(module, decl),
+            else => unreachable,
+        };
+    }
 
-    pub const Export = struct {
-        sym_index: ?u32 = null,
-    };
+    pub fn allocateDeclIndexes(base: *File, decl: *Module.Decl) !void {
+        try switch (base.tag) {
+            .Elf => @fieldParentPtr(Elf, "base", base).allocateDeclIndexes(decl),
+            .C => {
+                //TODO
```

Comment:
This should be just `return` right? What is there to do?

Reply:
Mostly, determine if there was anything to do :)